daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala

```diff
@@ -216,9 +213,13 @@ object Node {
       signatories: Set[Party],
       stakeholders: Set[Party],
       key: Option[KeyWithMaintainers[Val]],
-      override val byKey: Boolean // invariant (!byKey || exerciseResult.isDefined)
+      byKey: Option[Boolean] // None if unknown
   ) extends LeafOnlyNode[Cid, Val]
-      with NodeInfo.Fetch
+      with NodeInfo.Fetch {
+
+    require(byKey.forall(!_ || key.isDefined))
```

Comment:
It took me a moment to grasp this. Maybe this is simpler:
```suggestion
    require(byKey != Some(true) || key.isDefined)
```
Same below.

Reply:
But strictly less safe; you will not get a type error if `byKey`'s type changes, whereas the prior form compiles iff `byKey` has a `forall` method and its typearg has a `!` method that returns a value with a `||` method that returns the result expected by `forall`.

Reply:
Why would it not fail with a type error? Is the type of `!=` too general or would it use `!=` at the least common supertype of `byKey`'s type and `Option[Boolean]`?

Reply:
How about `!byKey.contains(true) || key.isDefined)`?

Reply:
@hurryabit The former; if you want any safety you must either explicitly type both arguments or use an alternative equality operator from a library.

Reply:
@S11001001 This is sad. 😿
@remyhaemmerle-da Let's go with @andreaslochbihler-da's suggestion then.